### PR TITLE
Testing C API

### PR DESCRIPTION
**Pull request by cwfitzgerald**
This PR adds a way of minimally testing the correctness of the C api. One `.c` file that includes the c api one `.cpp` which includes the c++ api. They each call bve_init, requiring proper linking and interop. 

CI builds and runs both of these to verify the includes are well formed, everything links correctly, and interop happens correctly.

Additionally, mitigations to the cbindgen/rustc bug https://github.com/rust-lang/rust/issues/68333 are included in bve-build.

There are a lot of commits because holy hell is compiling c++ annoying esp on ci, so just look at the resulting diff.

### CODE REVIEW

File: bve-build/src/main.rs [hunk 1]
```diff
@@ -45,58 +50,112 @@ use structopt::StructOpt;
...
+    let mut child = Command::new("cargo")
+        .arg("clean")
+        .spawn()
+        .expect("Unable to spawn cargo.");
+    assert!(child.wait().expect("Unable to wait for cargo.").success());
+}
+
+fn build(options: &Options) {
+    let mut args = if options.debug {
+        vec![String::from("build")]
```
Comment (DethRaid) [inline bve-build/src/main.rs]: Imagine using a language without string literals

Comment (cwfitzgerald) [reply bve-build/src/main.rs]: Imagine using a shuttup

File: bve-build/src/main.rs [hunk 2]
```diff
@@ -45,58 +50,112 @@ use structopt::StructOpt;
...
+        config.export.prefix = None;
+        *config.header.as_mut().expect("bve-native/cbindgen.toml needs a header") +=
+            "/* C++ API for BVE-Reborn high performance libraries. */";
+        config.trailer = Some(read_to_string("bve-native/include/bve_cpp.hpp").unwrap());
+        let result = cbindgen::Builder::new()
+            .with_crate("bve-native")
+            .with_config(config)
+            .generate();
+        match result {
+            Ok(bindings) => {
```
Comment (DethRaid) [inline bve-build/src/main.rs]: Weird that your `Ok` branch has {} but your `Err` one doesn't, but I guess code formatting is hard

File: bve-native/src/parse/mesh.rs [hunk 3]
```diff
@@ -9,10 +9,10 @@ use bve_derive::c_interface;
 use std::ffi::CStr;
 use std::ptr::null;
 
-pub use mesh::BlendMode as Blend_Mode;
-pub use mesh::FileType as File_Type;
-pub use mesh::Glow as Mesh_Glow;
-pub use mesh::GlowAttenuationMode as Mesh_Glow_Attenuation_Mode;
+pub use mesh::BlendMode;
```
Comment (DethRaid) [inline bve-native/src/parse/mesh.rs]: Why were these types aliased? Why do they no longer need to be aliased?

File: bve-native/tests/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    file(COPY "${CMAKE_SOURCE_DIR}/../../target/debug/bve_native.dll" DESTINATION ${CMAKE_BINARY_DIR}/Debug/)
+elseif(APPLE)
+    target_link_libraries(bve-native-c ${CMAKE_SOURCE_DIR}/../../target/debug/libbve_native.dylib)
+    target_link_libraries(bve-native-cpp ${CMAKE_SOURCE_DIR}/../../target/debug/libbve_native.dylib)
+elseif(UNIX)
+    target_link_libraries(bve-native-c ${CMAKE_SOURCE_DIR}/../../target/debug/libbve_native.so)
+    target_link_libraries(bve-native-cpp ${CMAKE_SOURCE_DIR}/../../target/debug/libbve_native.so)
+else()
+    message(FATAL_ERROR "Unsupported platform")
+endif()
```
Comment (DethRaid) [inline bve-native/tests/CMakeLists.txt]: Newline missing

File: bve-build/src/main.rs (hunk 2 above)
Comment (cwfitzgerald) [reply bve-build/src/main.rs]: Yeah I guess, I don't really notice things like that and try to not use parens if at all possible.

File: bve-native/src/parse/mesh.rs (hunk 3 above)
Comment (cwfitzgerald) [reply bve-native/src/parse/mesh.rs]: These type were already C compatible, so didn't need a manual wrapper. I was aliasing them to bring them into the naming convention of the C api. However cbindgen shat itself and couldn't find these types when aliased like this. I could newtype them (make a `#[repr(transparent)]` wrapper around them to bring it into line if you think it's worth it.

File: bve-native/tests/CMakeLists.txt (hunk 4 above)
Comment (cwfitzgerald) [reply bve-native/tests/CMakeLists.txt]: On it.

**Pull request merged**